modules/rcir.py

```diff
@@ -1,5 +1,216 @@
+        self.id: str = uuid4().hex
+        self.name: str = name
+        self.preferences_list: List[Candidate] = preferences_list
+
+
+class Election:
+    """
+    Election Class->
+
+    #TODO dubious terminology
+    @data winner: Candidate - Candidate who wins the election
+    @data majority: int - number of votes required to be considered a victor
+    @param voters: List[Voter] - list of all voters in the election
+    @param candidates: List[Candidate] - list of all candidates in the election
+
+    @method tabulate() -> None
+        * Iterate through the list of the candidates to find the latest
+        candidate in each voters' preference list that has not been eliminated.
+        Add a vote to this candidate
+    """
+
+    def __init__(self, voters: List[Voter], candidates: List[Candidate]) -> None:
+        """
+        constructor
+        @param voters - list of voters with their preference lists
+        @param candidates - list of all candidates with their statuses
+        """
+        self.voters: List[Voter] = voters
+        # make this a dict so it's easier to lookup candidates to add voters or
+        # eliminate candidates - this lookup is constant time
+        self.candidates: Dict[str, Candidate] = {x.id: x for x in candidates}
+        self.winner: Candidate = None
+        # majority count - 50% + 1
+        self.majority: int = int(len(self.voters) / 2) + 1
+
+    def get_polar_candidates(self: Election) -> List[str]:
+        # return the id's of the candidates with the lowest number of votes and
+        # highest number of votes
+        # there might be a better way to do this
+        # TODO optimize
+        candidates_and_votes: List[List[str, int]] = [
+            [candidate.id, candidate.votes] for candidate in self.candidates
+        ]
+        # sort this list by the number of votes
+        candidates_and_votes.sort(key=lambda x: x[1])
+
+        # find the number of candidates with the lowest number of votes
+        lowest_votes_candidate: Candidate = candidates_and_votes[0]
+        highest_votes_candidate: Candidate = candidates_and_votes[-1]
+
+        # return the required IDs
+        return (lowest_votes_candidate.id, highest_votes_candidate.id)
+
+    def add_vote_to_candidate(self: Election, candidate_id: str) -> None:
+        # add a vote to the candidate by ID
+        try:
+            candidate: Candidate = self.candidates[candidate_id]
+            candidate.addVote()
+        except KeyError:
+            print(f"Candidate with ID {candidate_id} does not exist")
+            exit()
+
+    def eliminate_candidate(self: Election, candidate_id: str) -> None:
+        # eliminate a candidate by id
+        try:
+            candidate: Candidate = self.candidates[candidate_id]
+            candidate.eliminate()
+        except KeyError:
+            print(f"Candidate with ID {candidate_id} does not exist")
+            exit()
```

Comment:
A better way (or at least I think it's better) to do this would be something like:

```suggestion
        if candidate_id in self.candidates
            candidate: Candidate = self.candidates[candidate_id]
            candidate.eliminate()
        else:
            print(f"Candidate with ID {candidate_id} does not exist")
            exit()
```

(I don't know if the formatting is correct, writing on phone xD)

Also, another suggestion I have is to use the logging module to log errors, warnings, info, etc. instead of using plain print statements.